apps/system/js/bluetooth_v2.js

```diff
@@ -0,0 +1,430 @@
+
+  /**
+   * Debug message.
+   *
+   * @type {Boolean} turn on/off the console log
+   */
+  onDebug: false,
+
+  /**
+   * Store a reference of the default adapter.
+   * If the adapter is not availble, no bluetooth operation could be
+   * executed.
+   *
+   * @private
+   */
+  _adapter: null,
+
+  /**
+   * Hold instance of bluetooth defaultAdapter change handler.
+   *
+   * @private
+   */
+  _btDefaultAdapterChangeHandler: null,
+
+  /**
+   * Hold instance of bluetooth adapter state change handler.
+   *
+   * @private
+   */
+  _btStateChangeHandler: null,
+
+  /**
+   * Store a reference for mozBluetooth.
+   *
+   * @private
+   */
+  _bluetooth: null,
+
+  /**
+   * State of Bluetooth default adapter.
+   *
+   * @private
+   */
+  _isEnabled: false,
+
+  /**
+   * Build-in Bluetooth profiles.
+   *
+   * @public
+   */
+  get Profiles() {
+    return {
+      HFP: 'hfp',   // Hands-Free Profile
+      OPP: 'opp',   // Object Push Profile
+      A2DP: 'a2dp', // A2DP status
+      SCO: 'sco'    // Synchronous Connection-Oriented
+    };
+  },
+
+  /**
+   * Set profile connect state.
+   *
+   * @public
+   * @param {String} profile profile id
+   * @param {Boolean} connected connect status
+   */
+  _setProfileConnected: function bt_setProfileConnected(profile, connected) {
+    var value = this['_' + profile + 'Connected'];
+    if (value !== connected) {
+      this['_' + profile + 'Connected'] = connected;
+
+      // Raise an event for the profile connection changes.
+      window.dispatchEvent(new CustomEvent('bluetoothprofileconnectionchange',
+      {
+        name: profile,
+        connected: connected
+      }));
+      // if (profile === 'opp' && this.transferIcon) {
+      //   this.transferIcon.update();
+      // }
+      // if (profile === 'a2dp' && this.headphoneIcon) {
+      //   this.headphoneIcon.update();
+      // }
+    }
+  },
+
+  /**
+   * Check if bluetooth profile is connected.
+   *
+   * @public
+   * @param {String} profile profile name
+   * @return {Boolean} connected state
+   */
+  isProfileConnected: function bt_isProfileConnected(profile) {
+    var isConnected = this['_' + profile + 'Connected'];
+    if (isConnected === undefined) {
+      return false;
+    } else {
+      return isConnected;
+    }
+  },
+
+  /**
+   * Initialize bluetooth module.
+   *
+   * @public
+   */
+  start: function bt_start() {
+    if (!window.navigator.mozSettings || !window.navigator.mozBluetooth) {
+      return;
+    }
+
+    this._bluetooth = window.navigator.mozBluetooth;
+
+    // listen states when the adapter is available
+    window.addEventListener('bluetooth-available',
+      this._registerListeners.bind(this));
+
+    // Generally bluetooth adapter is already available when start
+    // this module and no related event will be received.
+    // so we try to get the adapter right now
+    this._debug('init bluetooth adapter');
+    this._adapter = this._bluetooth.defaultAdapter;
```

Comment:
From this line to line 158: Some initialised work in this module. I would like to suggest having a function to it.